cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java

```diff
@@ -443,4 +500,147 @@ private long findProgramFinishTime(List<Message> messages) {
+      }
+
+      ArrayList<MetricValues> decoded = new ArrayList<>();
+
+      // Decode the messages into MetricValues.
+      String lastMessageId = null;
+      for (Message message : messages) {
+        lastMessageId = message.getId();
+        decodingOutputStream.reset(message.getPayload());
+        try {
+          decoded.add(reader.read(decoder, schema));
+        } catch (IOException e) {
+          LOG.warn("Failed to decode metric message.", e);
+        }
+      }
+      cachemap.clear();
+
+      // Aggregate the metrics.
+      ArrayList<MetricValues> aggregated = aggregateMetricValues(decoded);
+      ArrayList<Message> result = new ArrayList<>();
+
+      // Encode metrics back to messages.
+      for (MetricValues metricValues : aggregated) {
+        // Encode the metric value into a message.
+        encodeOutputStream.reset();
+        try {
+          recordWriter.encode(metricValues, encoder);
+          byte[] payload = encodeOutputStream.toByteArray();
+          // creating a final copy to access in inner class.
+          final String finalLastMessageId = lastMessageId;
+          result.add(new Message() {
+            @Override
+            public String getId() {
+              // TopicRelayers keep track of the last message ID
+              // to fetch messages in the next iteration.
+              // Message IDs are generated by the messaging service. Return the last message ID before aggregation
+              // every time.
+              return finalLastMessageId;
+            }
+
+            @Override
+            public byte[] getPayload() {
+              return payload;
+            }
+          });
+        } catch (IOException e) {
+          LOG.warn("Failed to encode metric to message payload.", e);
+        }
+      }
+
+      return result.iterator();
+    }
+
+    private ArrayList<MetricValues> aggregateMetricValues(ArrayList<MetricValues> metrics) {
+      ArrayList<MetricValues> result = new ArrayList<>();
+
+      // Group metrics with the same tags.
+      HashMap<Map<String, String>, ArrayList<MetricValues>> grouped = new HashMap<>();
+      for (MetricValues mv : metrics) {
+        if (mv.getMetrics().isEmpty()) {
+          continue;
+        }
+        grouped.putIfAbsent(mv.getTags(), new ArrayList<>());
+        grouped.get(mv.getTags()).add(mv);
+      }
+
+      // Aggregate metrics with the same tags and name.
+      for (Map.Entry<Map<String, String>, ArrayList<MetricValues>> iter : grouped.entrySet()) {
+        // Sort according to timestamp so that gauge values that were emitted at the end are retained.
+        ArrayList<MetricValues> sorted = iter.getValue();
+        sorted.sort((MetricValues m1, MetricValues m2) -> (int) (m1.getTimestamp() - m2.getTimestamp()));
+        HashMap<String, AggregatedMetricsEmitter> emitters = new HashMap<>();
+        ArrayList<MetricValue> aggregatedMetrics = new ArrayList<>();
+
+        for (MetricValues mvs : sorted) {
```

Comment:
Are we just aggregating everything within the same metrics batch? That doesn't sound right to me. If there is late arrival, we could be aggregating wrongly?

Reply:
Yes, that's the intention in the current code. The timestamp of the latest message is retained for all `MetricValues` with the same tag. So to the user, the metrics will show up at a later timestamp than when they were emitted. Since we fetch metrics every 2 seconds by default, the difference should be small in most cases, but it is possible to have bigger differences, for eg if there was a temporary network issue b/w CDAP and Hadoop cluster or b/w yarn containers.

To mitigate this we can use time windows of fixed size for aggregation. Do you think this is required? Do you have other approaches in mind to prevent this?

Reply:
Added a config to control the aggregation window size.